### PR TITLE
[REVIEW] Avoid redundant deepcopy in `cudf.from_pandas`

### PR DESCRIPTION
**Pull request by galipremsagar**
- [x] This PR removes a call to `df.set_index`, which internally does a deep copy of entire dataframe. This isn't necessary in `from_pandas` method since the data-frame is being constructed from scratch and doing this uncovered an issue with `base_mask` being set for columns that don't have nulls too. 
- [x] Fixes an issue with `StructColumn.from_arrow` where the null mask of the parent wasn't being propagated to child columns, this is resolved by calling into libcudf's `cudf::from_arrow` instead of manually constructing child columns at the python level.

### CODE REVIEW

File: python/cudf/cudf/core/column/column.py [hunk 1]
```diff
@@ -1988,10 +1988,10 @@ def as_column(
         data = Buffer(arbitrary._data.view("|u1"))
         data = build_column(data, dtype=cudf_dtype)
 
-        mask = arbitrary._mask
-        mask = bools_to_mask(as_column(mask).unary_operator("not"))
-
-        data = data.set_mask(mask)
+        pandas_mask = arbitrary._mask
+        if pandas_mask.any():
```
Comment (galipremsagar) [inline python/cudf/cudf/core/column/column.py:1992]: Can we avoid calling `.any()` to check if the pandas array has nulls or not?

Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py:1990]: An alternative that comes to my mind is converting to arrow without performing this mask introspection:

```
data = as_column(pa.Array.from_pandas(arbitrary), dtype=dtype)
```

Comment (shwina) [reply python/cudf/cudf/core/column/column.py:1990]: I'd benchmark to see if that's much faster?

**Review by shwina — APPROVED** [review]

File: python/cudf/cudf/core/column/column.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py:1990]: via pyarrow is a tiny bit faster:

```python
# mask introspection:

In [1]: import cudf

In [2]: import pandas as pd

In [3]: df = pd.DataFrame({"a":[10, 11, None]*10000000, "b":["a", "b", None]*10000000, "c":["abcdw", "weere", None]*10000000})

In [4]: %timeit gdf = cudf.DataFrame.from_pandas(df)
1.27 s ± 30.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: df = pd.DataFrame({"a":[10, 11, 12]*10000000, "b":["a", "b", "c"]*10000000, "c":["abcdw", "weere", "hig"]*10000000})

In [6]: %timeit gdf = cudf.DataFrame.from_pandas(df)
1.43 s ± 37.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)



# via pyarrow:

In [1]: import pandas as pd

In [2]: import cudf

In [3]: df = pd.DataFrame({"a":[10, 11, None]*10000000, "b":["a", "b", None]*10000000, "c":["abcdw", "weere", None]*10000000})

In [4]: %timeit gdf = cudf.DataFrame.from_pandas(df)
1.21 s ± 52.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: df = pd.DataFrame({"a":[10, 11, 12]*10000000, "b":["a", "b", "c"]*10000000, "c":["abcdw", "weere", "hig"]*10000000})

In [6]: %timeit gdf = cudf.DataFrame.from_pandas(df)
1.38 s ± 33.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py]: Switched to going via `pyarrow`

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**